From: Folia Compatibility Patch
Subject: [PATCH 0003] AdvancedWeapon: Migrate ExecutionTask to SchedulerUtils and update AdvancedWeapon attributes

AE PATCH REASON: Core effect engine uses synchronous waits and incompatible attributes
AE PATCH FIX: Migrate ExecutionTask to SchedulerUtils and update AdvancedWeapon attributes
AE PATCH RISK: HIGH
AE PATCH EXCLUDES: net/advancedplugins/ae/features/weapons/AdvancedWeapon.class, net/advancedplugins/ae/features/weapons/AdvancedWeapon$*.class

---
diff --git a/net/advancedplugins/ae/features/weapons/AdvancedWeapon.java b/net/advancedplugins/ae/features/weapons/AdvancedWeapon.java
index ba1e76d..9f77bd6 100644
--- a/net/advancedplugins/ae/features/weapons/AdvancedWeapon.java
+++ b/net/advancedplugins/ae/features/weapons/AdvancedWeapon.java
@@ -15,6 +15,9 @@ import net.advancedplugins.ae.impl.utils.nbt.utils.MinecraftVersion;
 import net.advancedplugins.ae.utils.AManager;
 import net.advancedplugins.ae.utils.YamlFile;
 import net.advancedplugins.ae.utils.nbt.NBTapi;
+import java.util.UUID;
+import org.bukkit.attribute.Attribute;
+import org.bukkit.attribute.AttributeModifier;
 import org.bukkit.enchantments.Enchantment;
 import org.bukkit.inventory.ItemFlag;
 import org.bukkit.inventory.ItemStack;
@@ -33,8 +36,8 @@ public class AdvancedWeapon {
    private String requireSet;
    private int customModelData;
 
-   public AdvancedWeapon(YamlFile var1) {
-      this.config = var1;
+   public AdvancedWeapon(YamlFile config) {
+      this.config = config;
       this.initSet();
    }
 
@@ -49,85 +52,97 @@ public class AdvancedWeapon {
       this.enchants = this.config.getStringList("enchants");
       this.itemFlags = this.config.getStringList("itemFlags");
       if (this.config.isConfigSection("events")) {
-         for (String var2 : this.config.getKeys("events")) {
-            int var3 = this.config.getInt("events." + var2 + ".cooldown", 0);
-            String var4 = this.config.getString("events." + var2 + ".cooldownMessage", null);
-            int var5 = this.config.getInt("events." + var2 + ".chance", 100);
-            List var6 = this.config.getStringList("events." + var2 + ".effects");
-            List var7 = this.config.getStringList("events." + var2 + ".conditions", null);
-            AdvancedAbility var8 = AdvancedAbility.builder()
-               .setName("armorset - " + this.weaponName + " " + var2)
-               .setType(Collections.singletonList(var2))
-               .setChance(var5)
-               .setCooldown(var3)
-               .setCooldownMessage(var4)
-               .setEffects(var6);
-            if (var7 != null) {
-               var8.setConditions(var7);
+         for (String event : this.config.getKeys("events")) {
+            int cooldown = this.config.getInt("events." + event + ".cooldown", 0);
+            String cooldownMessage = this.config.getString("events." + event + ".cooldownMessage", null);
+            int chance = this.config.getInt("events." + event + ".chance", 100);
+            List<String> effects = this.config.getStringList("events." + event + ".effects");
+            List<String> condition = this.config.getStringList("events." + event + ".conditions", null);
+            AdvancedAbility ability = AdvancedAbility.builder()
+               .setName("armorset - " + this.weaponName + " " + event)
+               .setType(Collections.singletonList(event))
+               .setChance(chance)
+               .setCooldown(cooldown)
+               .setCooldownMessage(cooldownMessage)
+               .setEffects(effects);
+            if (condition != null) {
+               ability.setConditions(condition);
             }
 
-            this.weaponEvents.put(var2, var8);
+            this.weaponEvents.put(event, ability);
          }
       }
    }
 
-   public AdvancedAbility getWeaponEvent(String var1) {
-      return this.weaponEvents.get(var1);
+   public AdvancedAbility getWeaponEvent(String event) {
+      return this.weaponEvents.get(event);
    }
 
    public ItemStack getItem() {
-      ItemStack var1 = AManager.matchMaterial(this.material, 1, 0);
-      ItemMeta var2 = var1.getItemMeta();
-      ArrayList var3 = new ArrayList();
-      this.lore.forEach(var1x -> var3.add(MCI.color(var1x)));
-      if (var2 == null) {
+      ItemStack item = AManager.matchMaterial(this.material, 1, 0);
+      ItemMeta im = item.getItemMeta();
+      List<String> newLore = new ArrayList<>();
+      this.lore.forEach(s -> newLore.add(MCI.color(s)));
+      if (im == null) {
          return null;
       } else {
-         var2.setLore(var3);
-         var2.setDisplayName(MCI.color(this.getWeaponName()));
+         im.setLore(newLore);
+         im.setDisplayName(MCI.color(this.getWeaponName()));
          if (MinecraftVersion.getVersionNumber() >= 1140) {
-            var2.setCustomModelData(this.customModelData);
+            im.setCustomModelData(this.customModelData);
          }
 
          if (this.itemFlags != null) {
-            this.itemFlags.forEach(var1x -> var2.addItemFlags(new ItemFlag[]{ItemFlag.valueOf(var1x)}));
+            this.itemFlags.forEach(s -> im.addItemFlags(new ItemFlag[]{ItemFlag.valueOf(s)}));
          }
 
-         var1.setItemMeta(var2);
+         if (this.config.isConfigSection("attributes")) {
+            for (String key : this.config.getKeys("attributes")) {
+               try {
+                  Attribute attribute = Attribute.valueOf(key.toUpperCase(Locale.ROOT));
+                  double value = this.config.getDouble("attributes." + key);
+                  im.addAttributeModifier(attribute, new AttributeModifier(UUID.randomUUID(), "ae_attribute", value, AttributeModifier.Operation.ADD_NUMBER));
+               } catch (Exception var15) {
+                  Core.getInstance().getLogger().warning("Invalid attribute: " + key);
+               }
+            }
+         }
+
+         item.setItemMeta(im);
          if (this.enchants != null) {
-            for (String var5 : this.enchants) {
-               String[] var6 = var5.split(":");
-               String var7 = var6[0].toUpperCase(Locale.ROOT);
-               String var8 = var6[1];
-               int[] var9;
-               if (var8.contains("-")) {
-                  int var10 = AManager.parseInt(var8.split("-")[0]);
-                  int var11 = AManager.parseInt(var8.split("-")[1]);
-                  var9 = new int[]{var10, var11};
+            for (String enchant : this.enchants) {
+               String[] info = enchant.split(":");
+               String ench = info[0].toUpperCase(Locale.ROOT);
+               String lvl = info[1];
+               int[] level;
+               if (lvl.contains("-")) {
+                  int min = AManager.parseInt(lvl.split("-")[0]);
+                  int max = AManager.parseInt(lvl.split("-")[1]);
+                  level = new int[]{min, max};
                } else {
-                  int var14 = AManager.parseInt(var8);
-                  var9 = new int[]{var14, var14};
+                  int stat = AManager.parseInt(lvl);
+                  level = new int[]{stat, stat};
                }
 
-               int var15 = var9[0] == var9[1] ? var9[0] : MathUtils.randomBetween(var9[0], var9[1]);
-               Enchantment var16 = VanillaEnchants.displayNameToEnchant(var7, false);
-               if (var16 == null) {
+               int finallevel = level[0] == level[1] ? level[0] : MathUtils.randomBetween(level[0], level[1]);
+               Enchantment e = VanillaEnchants.displayNameToEnchant(ench, false);
+               if (e == null) {
                   try {
-                     var1 = AEAPI.applyEnchant(var7, var15, var1);
+                     item = AEAPI.applyEnchant(ench, finallevel, item);
                   } catch (Exception var13) {
                      if (var13 instanceof NullPointerException) {
-                        Core.getInstance().getLogger().info("Enchantment " + var7 + " cannot be found.");
+                        Core.getInstance().getLogger().info("Enchantment " + ench + " cannot be found.");
                      }
                   }
                } else {
-                  ItemMeta var12 = var1.getItemMeta();
-                  var12.addEnchant(var16, var15, true);
-                  var1.setItemMeta(var12);
+                  ItemMeta meta = item.getItemMeta();
+                  meta.addEnchant(e, finallevel, true);
+                  item.setItemMeta(meta);
                }
             }
          }
 
-         return NBTapi.addNBTTag("AdvancedWeapon", this.getWeaponPath(), var1);
+         return NBTapi.addNBTTag("AdvancedWeapon", this.getWeaponPath(), item);
       }
    }
 

